src/models/ProjectNotification.ts

```diff
@@ -26,6 +27,96 @@ export class ProjectNotification {
         await this.notifyDanglingDataValues(recipients);
     }
 
+    async notifyOnDataReady(period: string, id: string) {
+        const res = await this.api.metadata
+            .get({
+                userRoles: {
+                    fields: {
+                        id: true,
+                        users: {
+                            email: true,
+                            id: true,
+                            userGroups: {
+                                id: true,
+                                name: true,
+                            },
+                        },
+                    },
+                    filter: { name: { in: baseConfig.userRoles.dataReviewer } },
+                },
+                dataSets: {
+                    fields: {
+                        id: true,
+                        userGroupAccesses: true,
+                        userAccesses: { id: true },
+                    },
+                    filter: { id: { in: [id] } },
+                },
+            })
+            .getData();
+
+        const { displayName: user, username } = this.currentUser.data;
+
+        const subject = i18n.t("{{username}} is requesting for data review", { username });
+        const pageLink = window.location.href;
+        const dataSetType = pageLink.includes("actual") ? "ACTUAL" : "TARGET";
+        const userAccessEmails = res.userRoles[0].users
+            .filter(user => {
+                return res.dataSets[0].userAccesses.some(userAccess => {
+                    return userAccess.id === user.id;
+                });
+            })
+            .map(user => user.email);
+        const userGroupEmails = res.userRoles[0].users
+            .map(user => {
+                const a = user.userGroups.filter(userGroup => {
+                    return res.dataSets[0].userGroupAccesses.some(userGroupAccess => {
+                        return userGroupAccess.id === userGroup.id;
+                    });
+                });
+                return a;
+            })
+            .map((e, i) => {
+                if (e.length !== 0) {
+                    return i;
+                } else {
+                    return null;
+                }
+            })
+            .filter(element => element && element >= 0)
+            .map(e => {
+                return res.userRoles[0].users[e ? e : 0];
+            })
```

Comment:
This map+map+filter+map does not look right. I think you want first a filter (with two `some` inside) + map. Similar to the previous code.